### PR TITLE
feat(unify): toggle specs list

### PR DESCRIPTION
**Pull request by mapsandapps**
Closes https://cypress-io.atlassian.net/browse/UNIFY-551

### User facing changelog

The inline specs list in the runner can be opened/closed by means of the Specs button in the reporter, as well as the shortcut key `f`.

### Additional details

This PR does some other things (paired with @lmiller1990 on some of this):

- uses the Runner UI Store for Vue state of the specs list as open/closed, but also to initialize the reporter with the correct value from gql once the localPreferences have been loaded
- fixes a missing hookup for the `autoScrollingEnabled` flag using the same flow as the `isSpecsListOpen` state
- fixes a tiny bug with saving `autoScrollingEnabled` when toggled via keyboard shortcut
- there's some CSS stuff from @mapsandapps that seems like it's fine to stay in - Shawn let me know if this is no longer needed?

### How has the user experience changed?

https://user-images.githubusercontent.com/8340719/144281985-10a0a97b-934d-45e8-9d1e-59a74b3b58f1.mov

### How to test

- Run `yarn dev` for a project with specs and go to a spec.
- Toggle specs list with `f` key (when not focused in a text field or the AUT)
- Toggle specs list with the Specs button in the reporter
- Search for something in the specs list and toggle - see that specs list internal state is maintained
- See that open/closed state is maintained after page refresh, navigation, and full stop and restart of Cypress

### PR Tasks

- [x] Have tests been added/updated?
- [na] Has the original issue or this PR been tagged with a release in ZenHub? 
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [na] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

**Comment by mapsandapps** [comment]
@lmiller1990 @marktnoonan following up on conversation from #18921 

i got one-way communication working, but we do need to communicate back the other way as well, so the button will know if the spec list is open or not and display the correct icon. i didn't have time to start digging into that today, but i should have time to start looking at it tomorrow.

File: packages/reporter/cypress/integration/unit/events_spec.ts [hunk 1]
```diff
@@ -363,16 +363,13 @@ describe('events', () => {
       expect(runner.emit).to.have.been.calledWith('runner:unpin:snapshot', 'command id')
     })
 
-    it('emits focus:tests on focus:tests', () => {
-      events.emit('focus:tests')
-      expect(runner.emit).to.have.been.calledWith('focus:tests')
-    })
-
```
Comment (marktnoonan) [inline packages/reporter/cypress/integration/unit/events_spec.ts:370]: This test change is pretty representative of what's going on in this PR - we no longer "focus the launchpad that's showing the specs list" but instead, change it as a setting that's part of reporter state, just like `autoScrollingEnabled` has worked until now.

File: packages/reporter/src/lib/shortcuts.ts [hunk 2]
```diff
@@ -23,14 +23,19 @@ class Shortcuts {
...
+      })()
+
         break
       case 'c': events.emit('resume')
         break
       case 'n': events.emit('next')
         break
       case 'a': action('set:scrolling', () => {
         appState.setAutoScrolling(!appState.autoScrollingEnabled)
+        events.emit('save:state')
```
Comment (marktnoonan) [inline packages/reporter/src/lib/shortcuts.ts:38]: This fixes a bug where this value was not getting saved when the shortcut key was used.

File: packages/server/lib/project-base.ts [hunk 3]
```diff
@@ -611,7 +611,9 @@ export class ProjectBase<TServer extends Server> extends EE {
       onReloadBrowser: options.onReloadBrowser,
       onFocusTests: options.onFocusTests,
       onSpecChanged: options.onSpecChanged,
-      onSavedStateChanged: (state: any) => this.saveState(state),
+      onSavedStateChanged: (state: any) => {
```
Comment (ZachJW34) [inline packages/server/lib/project-base.ts:614]: This change can be undone

**Review by ZachJW34 — COMMENTED** [review]
Tested, works great! Just a few minor comments.

We should change the shortcuts to properly reflect what's happening. `f` says it toggles between specs whereas `command + b` says it toggles the list.

Screen Shot 2021-12-01 at 4 05 15 PM

File: packages/reporter/src/header/header.tsx [hunk 4]
```diff
@@ -20,11 +21,20 @@ export interface ReporterHeaderProps {
 
 const Header = observer(({ appState, events = defaultEvents, statsStore }: ReporterHeaderProps) => (
   <header>
-    <Tooltip placement='bottom' title={<p>View All Tests <span className='kbd'>F</span></p>} wrapperClassName='focus-tests' className='cy-tooltip'>
-      <button onClick={() => events.emit('focus:tests')}>
-        <MenuExpandRightIcon />
+    <Tooltip placement='bottom' title={<p>View All Specs <span className='kbd'>F</span></p>} wrapperClassName='toggle-specs-wrapper' className='cy-tooltip'>
```
Comment (mapsandapps) [inline packages/reporter/src/header/header.tsx:24]: We discussed this already, but mentioning it here too: Please change the tooltip to toggle between “Expand Specs List” and “Collapse Specs List”. Sorry I missed this when I was doing the initial work for this ticket!

File: packages/reporter/cypress/integration/shortcuts_spec.ts [hunk 5]
```diff
@@ -73,13 +73,24 @@ describe('shortcuts', function () {
       })
     })
 
-    it('focuses on specs', () => {
+    it('toggles the specs list', () => {
+      // 1. save:state should be emitted to preserve the change in state
+      // 2. the reporter appSate should be also be updated with the new value,
+      // checking the aria-expanded state of the button confirms thos
```
[the remark this answers is not in the document]
Comment (JessicaSachs) [reply packages/reporter/cypress/integration/shortcuts_spec.ts:79]: Please install the Code Spell plugin :-)

File: packages/app/src/runner/SpecRunner.vue [hunk 6]
```diff
@@ -97,12 +106,27 @@ mutation OpenFileInIDE ($input: FileDetailsInput!) {
...
+  activeSpec: BaseSpec
+}>()
+
 const eventManager = getEventManager()
 
 const autStore = useAutStore()
 const screenshotStore = useScreenshotStore()
 const runnerUiStore = useRunnerUiStore()
 
+runnerUiStore.setAutoScrollingEnabled(props.gql.localSettings.preferences.autoScrollingEnabled ?? true)
```
Comment (lmiller1990) [inline packages/app/src/runner/SpecRunner.vue:125]: I foresee a bunch of default values coming in the future, we might want a `setDefaults` function at some point but prob fine for now. Good to keep in mind if this pattern of setting defaults expands.

File: packages/app/src/runner/SpecRunner.vue [hunk 7]
```diff
@@ -186,6 +205,12 @@ onMounted(() => {
   eventManager.on('after:screenshot', () => {
     screenshotStore.setScreenshotting(false)
   })
+
+  eventManager.on('save:app:state', (state) => {
+    runnerUiStore.setIsSpecsListOpen(state.isSpecsListOpen)
```
Comment (lmiller1990) [inline packages/app/src/runner/SpecRunner.vue:210]: Hmm now I see this I wonder if we want a single `updateState` in the `runnerUiStore` that would more or less 1:1 map to `setPreferences.executeMutation`. Thoughts? Consider we have many more things to come like appX, appY, paneWidth etc - we likely want to have some pattern to consistently update the local Pinia store, as well at GraphQL.

Comment (lmiller1990) [reply packages/app/src/runner/SpecRunner.vue:210]: Eg, I wonder if we want to write some kind of abstraction like an `usePreferences` composables, that does something like:

```ts
function usePreferences (preference: 'autoScrollingEnabled' | ... others ..., value: any) {

  function update(preference: 'autoScrollingEnabled' | ... others ..., value: any) {
    runnerUiStore.setPreference(preference, value)
    setPreferences.executeMutation({ value: JSON.stringify({ [preference]: value ) })
  }

  return {
    update
  }
}
```

What do you think? Then we just do

```ts
const prefs = usePreferences()
prefs.update('autoScrollingEnabled', false)
```

And everything will stay in sync. We could strongly type this; the work has already been done in `packages/types/src/preferences.ts`.

**Review by lmiller1990 — COMMENTED** [review]
Code seems fine but I see this abstraction with the graphql preferences and local state getting out of sync and unmanageable really quickly, what do you think of a quick abstraction to wrap it all up? See here: https://github.com/cypress-io/cypress/pull/18991/files#r761052458

File: packages/reporter/cypress/integration/shortcuts_spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (marktnoonan) [reply packages/reporter/cypress/integration/shortcuts_spec.ts:79]: Done and done :)

**Comment by marktnoonan** [comment]
> We should change the shortcuts to properly reflect what's happening. `f` says it toggles between specs whereas `command + b` says it toggles the list.

Good point @ZachJW34. And it's not totally clear which to remove. `Cmd+b` is the specs list toggle from the component test runner, and `f` is the old way of focusing the spec list in the launchpad. Possibly we should support both if people have adjusted to using them, even though they have converged to do the same thing.

I think there is a bit more work to do on the keyboard shortcuts popup in general, it's actually missing several as well. I'll update the text for `f` and make a ticket to revisit the overall shortcuts if we don't already have that.

File: packages/app/cypress/e2e/integration/navigation.spec.ts [hunk 8]
```diff
@@ -10,6 +10,7 @@ describe('Navigation', () => {
     cy.intercept('mutation-ExternalLink_OpenExternal', { 'data': { 'openExternal': true } }).as('OpenExternal')
     cy.contains('button', defaultMessages.topNav.docsMenu.docsHeading).click()
     cy.contains('a', defaultMessages.topNav.docsMenu.firstTest).click()
+
```
Comment (elevatebart) [inline packages/app/cypress/e2e/integration/navigation.spec.ts:13]: Was this changed on purpose ?

File: packages/app/src/navigation/KeyboardBindingsModal.spec.tsx [hunk 9]
```diff
@@ -1,9 +1,16 @@
 import KeyboardBindingsModal from './KeyboardBindingsModal.vue'
+import { defaultMessages } from '@cy/i18n'
 
 describe('SidebarNavigationRow', () => {
```
Comment (elevatebart) [inline packages/app/src/navigation/KeyboardBindingsModal.spec.tsx:4]: This one would be my fault...

```suggestion
describe('KeyboardBindingsModal', () => {
```

File: packages/app/src/runner/SpecRunner.vue [hunk 10]
```diff
@@ -29,7 +34,7 @@
 
     <div
       ref="runnerPane"
-      class="relative w-full"
+      class="w-full relative"
     >
       <HideDuringScreenshot class="bg-white p-4">
```
Comment (elevatebart) [inline packages/app/src/runner/SpecRunner.vue:39]:
```suggestion
      <HideDuringScreenshot class="bg-white p-16px">
```

File: packages/app/src/runner/SpecRunner.vue [hunk 11]
```diff
@@ -41,11 +46,11 @@
       </HideDuringScreenshot>
 
       <RemoveClassesDuringScreenshotting
-        class="flex justify-center bg-gray-100 h-full p-4"
+        class="flex h-full bg-gray-100 p-4 justify-center"
```
Comment (elevatebart) [inline packages/app/src/runner/SpecRunner.vue:49]:
```suggestion
        class="flex h-full bg-gray-100 p-16px justify-center"
```

**Review by elevatebart — CHANGES_REQUESTED** [review]
Some changes, I am going to test it locally

File: packages/app/cypress/e2e/integration/navigation.spec.ts (hunk 8 above)
Comment (marktnoonan) [reply packages/app/cypress/e2e/integration/navigation.spec.ts]: Nope!

**Review by JessicaSachs — APPROVED** [review]
tested manually.

**Pull request merged**